base/secretbuffer.jl

```diff
@@ -0,0 +1,160 @@
+"""
+    Base.SecretBuffer()
+
+An IOBuffer-like object where the contents will be securely wiped when garbage collected.
+
+It is considered best practice to wipe the buffer using `Base.shred!(::SecretBuffer)` as
+soon as the secure data are no longer required. When initializing with existing data, the
+`SecretBuffer!` method is highly recommended to securely zero the passed argument. Avoid
+initializing with and converting to `String`s as they are unable to be securely zeroed.
+
+# Examples
+```jldoctest
+julia> s = Base.SecretBuffer()
+SecretBuffer("*******")
+
+julia> write(s, 's', 'e', 'c', 'r', 'e', 't')
+6
+
+julia> seek(s, 0); Char(read(s, UInt8))
+'s': ASCII/Unicode U+0073 (category Ll: Letter, lowercase)
+
+julia> Base.shred!(s)
+SecretBuffer("*******")
+
+julia> eof(s)
+true
+```
+"""
+mutable struct SecretBuffer <: IO
+    data::Vector{UInt8}
+    size::Int
+    ptr::Int
+
+    function SecretBuffer(; sizehint=128)
+        s = new(Vector{UInt8}(undef, sizehint), 0, 1)
+        finalizer(final_shred!, s)
+        return s
+    end
+end
+
+convert(::Type{SecretBuffer}, s::AbstractString) = SecretBuffer(String(s))
+SecretBuffer(str::AbstractString) = SecretBuffer(String(str))
+function SecretBuffer(str::String)
```

Comment:
I think it bears creating a docstring for the `::String` constructor that mentions that the string will not be zeroed out.